File: MercaderSG/Gestión/AltaLocalidad.vb
Imports Entidades
Imports Excepciones
Imports Negocios
Public Class AltaLocalidad

    Private Sub AltaLocalidad_Load(sender As System.Object, e As System.EventArgs) Handles MyBase.Load
        AplicarIdioma()
        CargarTT()
        CargarFoco(LocalidadGB)

        ProvinciaCMB.DataSource = LocalidadRN.CargarProvincia()
        ProvinciaCMB.DisplayMember = "Descripcion"
        ProvinciaCMB.ValueMember = "CodProvincia"
    End Sub

    Private Sub AceptarBtn_Click(sender As System.Object, e As System.EventArgs) Handles AceptarBtn.Click
        If Not ConsistenciaDatos() Then
            Exit Sub
        End If

        Dim Localidad As New LocalidadEN
        Localidad.Descripcion = DescripcionTxt.Text
        Localidad.CodigoPostal = CodPostalTxt.Text
        Localidad.CodProvincia = ProvinciaCMB.SelectedValue

        Try
            LocalidadRN.AltaLocalidad(Localidad)
        Catch ex As WarningException
            MessageBox.Show(ex.Message, My.Resources.ArchivoIdioma.MsgBoxAdvertencia, MessageBoxButtons.OK, MessageBoxIcon.Warning)
            DescripcionTxt.Clear()
        Catch ex As InformationException
            MessageBox.Show(ex.Message, My.Resources.ArchivoIdioma.MsgBoxInformacion, MessageBoxButtons.OK, MessageBoxIcon.Information)
            DialogResult = Windows.Forms.DialogResult.OK
        End Try
    End Sub


#Region "Validaciones"

    Public Function ConsistenciaDatos() As Boolean
        Dim Resultado As Boolean = True

        If String.IsNullOrEmpty(DescripcionTxt.Text) Then
            ErrorP.SetError(DescripcionTxt, My.Resources.ArchivoIdioma.CampoVacio)
            Resultado = False
        Else
            ErrorP.SetError(DescripcionTxt, "")
        End If

        Select Case DescripcionTxt.Text.Length
            Case Is > 50
                ErrorP.SetError(DescripcionTxt, My.Resources.ArchivoIdioma.Contener50Carac)
                Resultado = False
            Case 1 To 50
                ErrorP.SetError(DescripcionTxt, "")
        End Select

        If String.IsNullOrEmpty(CodPostalTxt.Text) Then
            ErrorP.SetError(CodPostalTxt, My.Resources.ArchivoIdioma.CampoVacio)
            Resultado = False
        Else
            ErrorP.SetError(CodPostalTxt, "")
        End If

        Select Case CodPostalTxt.Text.Length
            Case Is > 10
                ErrorP.SetError(CodPostalTxt, My.Resources.ArchivoIdioma.Contener10Carac)
                Resultado = False
            Case 1 To 10
                ErrorP.SetError(CodPostalTxt, "")
        End Select

        Return Resultado
    End Function

    Private Sub DescripcionTxt_Validated(sender As System.Object, e As System.EventArgs) Handles DescripcionTxt.Validated
        DescripcionTxt.Text = UCase(Mid(DescripcionTxt.Text, 1, 1)) & LCase(Mid(DescripcionTxt.Text, 2))
    End Sub

    Private Sub DescripcionTxt_KeyPress(sender As System.Object, e As System.Windows.Forms.KeyPressEventArgs) Handles DescripcionTxt.KeyPress
        Dim CaracteresPermitidos As String = "áéúíóÁÉÍÓÚQWERTYUIOPASDFGHJKLÑZXCVBNMqwertyuiopasdfghjklñzxcvbnm " & Convert.ToChar(8)
        Dim c As Char = e.KeyChar
        If Not CaracteresPermitidos.Contains(c) Then
            e.Handled = True
        End If
    End Sub

    Private Sub CodPostalTxt_KeyPress(sender As System.Object, e As System.Windows.Forms.KeyPressEventArgs) Handles CodPostalTxt.KeyPress
        Dim CaracteresPermitidos As String = "1234567890" & Convert.ToChar(8)
        Dim c As Char = e.KeyChar
        If Not CaracteresPermitidos.Contains(c) Then
            e.Handled = True
        End If
    End Sub

    Private Sub DescripcionTxt_TextChanged(sender As System.Object, e As System.EventArgs) Handles DescripcionTxt.TextChanged
        ErrorP.SetError(DescripcionTxt, "")
    End Sub

    Private Sub CodPostalTxt_TextChanged(sender As System.Object, e As System.EventArgs) Handles CodPostalTxt.TextChanged
        ErrorP.SetError(DescripcionTxt, "")
    End Sub

    Private Sub AltaLocalidad_KeyDown(sender As System.Object, e As System.Windows.Forms.KeyEventArgs) Handles MyBase.KeyDown
        If e.KeyCode = Keys.Escape Then
            Me.Close()
        End If
    End Sub
#End Region

#Region "Mejoras"
    Private Sub TieneFoco(ByVal sender As System.Object, ByVal e As System.EventArgs)
        Dim MiTextBox As TextBox = DirectCast(sender, TextBox)
        MiTextBox.BackColor = Color.WhiteSmoke
    End Sub

    Private Sub PierdeFoco(ByVal sender As Object, ByVal e As System.EventArgs)
        Dim MiTextBox As TextBox = DirectCast(sender, TextBox)
        MiTextBox.BackColor = Color.White
    End Sub

    Public Sub CargarFoco(Grupo As Windows.Forms.GroupBox)
        Dim Ctrl As Control
        For Each Ctrl In Grupo.Controls
            If (TypeOf (Ctrl) Is System.Windows.Forms.TextBox) Then
                Dim MiTextBox As TextBox = DirectCast(Ctrl, TextBox)
                If MiTextBox.ReadOnly = False Then
                    AddHandler MiTextBox.GotFocus, AddressOf TieneFoco
                    AddHandler MiTextBox.LostFocus, AddressOf PierdeFoco
                End If
            End If
        Next
    End Sub
#End Region

#Region "Idioma y TT"

    Public Sub AplicarIdioma()
        Me.Text = My.Resources.ArchivoIdioma.AltaLocalidadFrm

        LocalidadGB.Text = My.Resources.ArchivoIdioma.LocalidadLbl
        DescripcionLbl.Text = My.Resources.ArchivoIdioma.DescripcionLbl
        CodPostalLbl.Text = My.Resources.ArchivoIdioma.CodPostalLbl
        ProvinciaLbl.Text = My.Resources.ArchivoIdioma.ProvinciaLbl

        AceptarBtn.Text = My.Resources.ArchivoIdioma.AceptarBtn
        CancelarBtn.Text = My.Resources.ArchivoIdioma.CancelarBtn
    End Sub

    Private Sub CargarTT()
        ControlesTP.SetToolTip(DescripcionTxt, My.Resources.ArchivoIdioma.TTDescripcionLoc)
        ControlesTP.SetToolTip(CodPostalTxt, My.Resources.ArchivoIdioma.TTCodPostalLoc)
        ControlesTP.SetToolTip(ProvinciaCMB, My.Resources.ArchivoIdioma.TTProvinciaLoc)
        ControlesTP.SetToolTip(AceptarBtn, My.Resources.ArchivoIdioma.TTAceptarAltaLoc)
        ControlesTP.SetToolTip(CancelarBtn, My.Resources.ArchivoIdioma.TTCancelarBtn)
    End Sub

#End Region
    
    Private Sub CancelarBtn_Click(sender As System.Object, e As System.EventArgs) Handles CancelarBtn.Click
        Me.Close()
    End Sub
End Class

File: MercaderSG/Gestión/AltaLocalidad.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using Entidades;
using Excepciones;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using Negocios;

namespace MercaderSG
{
    public partial class AltaLocalidad
    {
        public AltaLocalidad()
        {
            InitializeComponent();
            _CodPostalTxt.Name = "CodPostalTxt";
            _DescripcionTxt.Name = "DescripcionTxt";
            _AceptarBtn.Name = "AceptarBtn";
            _CancelarBtn.Name = "CancelarBtn";
        }

        private void AltaLocalidad_Load(object sender, EventArgs e)
        {
            AplicarIdioma();
            CargarTT();
            CargarFoco(LocalidadGB);
            ProvinciaCMB.DataSource = LocalidadRN.CargarProvincia();
            ProvinciaCMB.DisplayMember = "Descripcion";
            ProvinciaCMB.ValueMember = "CodProvincia";
        }

        private void AceptarBtn_Click(object sender, EventArgs e)
        {
            if (!ConsistenciaDatos())
            {
                return;
            }

            var Localidad = new LocalidadEN();
            Localidad.Descripcion = DescripcionTxt.Text;
            Localidad.CodigoPostal = CodPostalTxt.Text;
            Localidad.CodProvincia = Conversions.ToInteger(ProvinciaCMB.SelectedValue);
            try
            {
                LocalidadRN.AltaLocalidad(Localidad);
            }
            catch (WarningException ex)
            {
                MessageBox.Show(ex.Message, My.Resources.ArchivoIdioma.MsgBoxAdvertencia, MessageBoxButtons.OK, MessageBoxIcon.Warning);
                DescripcionTxt.Clear();
            }
            catch (InformationException ex)
            {
                MessageBox.Show(ex.Message, My.Resources.ArchivoIdioma.MsgBoxInformacion, MessageBoxButtons.OK, MessageBoxIcon.Information);
                DialogResult = DialogResult.OK;
            }
        }


        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        public bool ConsistenciaDatos()
        {
            bool Resultado = true;
            if (string.IsNullOrEmpty(DescripcionTxt.Text))
            {
                ErrorP.SetError(DescripcionTxt, My.Resources.ArchivoIdioma.CampoVacio);
                Resultado = false;
            }
            else
            {
                ErrorP.SetError(DescripcionTxt, "");
            }

            switch (DescripcionTxt.Text.Length)
            {
                case var @case when @case > 50:
                    {
                        ErrorP.SetError(DescripcionTxt, My.Resources.ArchivoIdioma.Contener50Carac);
                        Resultado = false;
                        break;
                    }

                case var case1 when 1 <= case1 && case1 <= 50:
                    {
                        ErrorP.SetError(DescripcionTxt, "");
                        break;
                    }
            }

            if (string.IsNullOrEmpty(CodPostalTxt.Text))
            {
                ErrorP.SetError(CodPostalTxt, My.Resources.ArchivoIdioma.CampoVacio);
                Resultado = false;
            }
            else
            {
                ErrorP.SetError(CodPostalTxt, "");
            }

            switch (CodPostalTxt.Text.Length)
            {
                case var case2 when case2 > 10:
                    {
                        ErrorP.SetError(CodPostalTxt, My.Resources.ArchivoIdioma.Contener10Carac);
                        Resultado = false;
                        break;
                    }

                case var case3 when 1 <= case3 && case3 <= 10:
                    {
                        ErrorP.SetError(CodPostalTxt, "");
                        break;
                    }
            }

            return Resultado;
        }

        private void DescripcionTxt_Validated(object sender, EventArgs e)
        {
            DescripcionTxt.Text = Strings.UCase(Strings.Mid(DescripcionTxt.Text, 1, 1)) + Strings.LCase(Strings.Mid(DescripcionTxt.Text, 2));
        }

        private void DescripcionTxt_KeyPress(object sender, KeyPressEventArgs e)
        {
            string CaracteresPermitidos = "áéúíóÁÉÍÓÚQWERTYUIOPASDFGHJKLÑZXCVBNMqwertyuiopasdfghjklñzxcvbnm " + Convert.ToChar(8);
            char c = e.KeyChar;
            if (!CaracteresPermitidos.Contains(Conversions.ToString(c)))
            {
                e.Handled = true;
            }
        }

        private void CodPostalTxt_KeyPress(object sender, KeyPressEventArgs e)
        {
            string CaracteresPermitidos = "1234567890" + Convert.ToChar(8);
            char c = e.KeyChar;
            if (!CaracteresPermitidos.Contains(Conversions.ToString(c)))
            {
                e.Handled = true;
            }
        }

        private void DescripcionTxt_TextChanged(object sender, EventArgs e)
        {
            ErrorP.SetError(DescripcionTxt, "");
        }

        private void CodPostalTxt_TextChanged(object sender, EventArgs e)
        {
            ErrorP.SetError(DescripcionTxt, "");
        }

        private void AltaLocalidad_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Escape)
            {
                Close();
            }
        }
        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        private void TieneFoco(object sender, EventArgs e)
        {
            TextBox MiTextBox = (TextBox)sender;
            MiTextBox.BackColor = Color.WhiteSmoke;
        }

        private void PierdeFoco(object sender, EventArgs e)
        {
            TextBox MiTextBox = (TextBox)sender;
            MiTextBox.BackColor = Color.White;
        }

        public void CargarFoco(GroupBox Grupo)
        {
            foreach (Control Ctrl in Grupo.Controls)
            {
                if (Ctrl is TextBox)
                {
                    TextBox MiTextBox = (TextBox)Ctrl;
                    if (MiTextBox.ReadOnly == false)
                    {
                        MiTextBox.GotFocus += TieneFoco;
                        MiTextBox.LostFocus += PierdeFoco;
                    }
                }
            }
        }
        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        public void AplicarIdioma()
        {
            Text = My.Resources.ArchivoIdioma.AltaLocalidadFrm;
            LocalidadGB.Text = My.Resources.ArchivoIdioma.LocalidadLbl;
            DescripcionLbl.Text = My.Resources.ArchivoIdioma.DescripcionLbl;
            CodPostalLbl.Text = My.Resources.ArchivoIdioma.CodPostalLbl;
            ProvinciaLbl.Text = My.Resources.ArchivoIdioma.ProvinciaLbl;
            AceptarBtn.Text = My.Resources.ArchivoIdioma.AceptarBtn;
            CancelarBtn.Text = My.Resources.ArchivoIdioma.CancelarBtn;
        }

        private void CargarTT()
        {
            ControlesTP.SetToolTip(DescripcionTxt, My.Resources.ArchivoIdioma.TTDescripcionLoc);
            ControlesTP.SetToolTip(CodPostalTxt, My.Resources.ArchivoIdioma.TTCodPostalLoc);
            ControlesTP.SetToolTip(ProvinciaCMB, My.Resources.ArchivoIdioma.TTProvinciaLoc);
            ControlesTP.SetToolTip(AceptarBtn, My.Resources.ArchivoIdioma.TTAceptarAltaLoc);
            ControlesTP.SetToolTip(CancelarBtn, My.Resources.ArchivoIdioma.TTCancelarBtn);
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        private void CancelarBtn_Click(object sender, EventArgs e)
        {
            Close();
        }
    }
}